service/history/historyReplicator.go

```diff
@@ -265,155 +344,28 @@ func (r *historyReplicator) ApplyReplicationTask(context *workflowExecutionConte
-		}
-		setTaskVersion(msBuilder.GetCurrentVersion(), sBuilder.transferTasks, sBuilder.timerTasks)
-
-		createWorkflow := func(isBrandNew bool, prevRunID string) (string, error) {
-			_, err = r.shard.CreateWorkflowExecution(&persistence.CreateWorkflowExecutionRequest{
-				RequestID:                   requestID,
-				DomainID:                    domainID,
-				Execution:                   execution,
-				ParentDomainID:              parentDomainID,
-				ParentExecution:             parentExecution,
-				InitiatedID:                 initiatedID,
-				TaskList:                    executionInfo.TaskList,
-				WorkflowTypeName:            executionInfo.WorkflowTypeName,
-				WorkflowTimeout:             executionInfo.WorkflowTimeout,
-				DecisionTimeoutValue:        executionInfo.DecisionTimeoutValue,
-				ExecutionContext:            nil,
-				NextEventID:                 msBuilder.GetNextEventID(),
-				LastProcessedEvent:          common.EmptyEventID,
-				TransferTasks:               sBuilder.transferTasks,
-				DecisionVersion:             decisionVersionID,
-				DecisionScheduleID:          decisionScheduleID,
-				DecisionStartedID:           decisionStartID,
-				DecisionStartToCloseTimeout: decisionTimeout,
-				TimerTasks:                  sBuilder.timerTasks,
-				ContinueAsNew:               !isBrandNew,
-				PreviousRunID:               prevRunID,
-				ReplicationState:            replicationState,
-			})
-
-			if err != nil {
-				return "", err
-			}
-
-			return execution.GetRunId(), nil
-		}
-
-		// TODO
-		// The failover version checking && overwriting should be performed here: #675
-		// TODO
-
-		// try to create the workflow execution
-		isBrandNew := true
-		_, err = createWorkflow(isBrandNew, "")
-		// if err still non nil, see if retry
-		if errExist, ok := err.(*persistence.WorkflowExecutionAlreadyStartedError); ok {
-			prevRunID := errExist.RunID
-			prevState := errExist.State
-
-			// Check for duplicate processing of StartWorkflowExecution replication task
-			if prevRunID == execution.GetRunId() {
-				r.logger.Infof("Dropping stale replication task for start event.  WorkflowID: %v, RunID: %v, Version: %v",
-					execution.GetWorkflowId(), execution.GetRunId(), request.GetVersion())
-				return nil
-			}
-
-			// Some other workflow is running, for now let's keep on retrying this replication event by an error
-			// to wait for current run to finish so this event could be applied.
-			// TODO: We also need to deal with conflict resolution when workflow with same ID is started on 2 different
-			// clusters.
-			if prevState != persistence.WorkflowStateCompleted {
-				return err
-			}
-
-			// if the existing workflow is completed, ignore the worklow ID reuse policy
-			// since the policy should be applied by the active cluster,
-			// standby cluster should apply this event without question.
-			isBrandNew = false
-			_, err = createWorkflow(isBrandNew, errExist.RunID)
-		}
-
+		requestID := uuid.New()
```

Comment:
Can we move all this logic into ApplyStartEvent?  We can simplify this logic to deal with only other events and have an assertion that it should not be called for WorkflowExecutionStartedEvent.

Reply:
actually, i think we should keep this logic like this.

all other functions, other than ApplyReplicationTask should prepare the mutable state.
ApplyReplicationTask should apply events to mutable state. and accordingly, create workflow or update workflow.
this can be useful if say, we want to apply events from a remote cluster to recover from a disaster, or simple reply a workflow using existing events for debugging reason